// react-menu-list/js/events/MenuEvent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var MenuEvent =
/*#__PURE__*/
function () {
  function MenuEvent(type) {
    (0, _classCallCheck2["default"])(this, MenuEvent);
    (0, _defineProperty2["default"])(this, "type", void 0);
    (0, _defineProperty2["default"])(this, "cancelBubble", false);
    (0, _defineProperty2["default"])(this, "defaultPrevented", false);
    this.type = type;
  }

  (0, _createClass2["default"])(MenuEvent, [{
    key: "stopPropagation",
    value: function stopPropagation() {
      this.cancelBubble = true;
    }
  }, {
    key: "preventDefault",
    value: function preventDefault() {
      this.defaultPrevented = true;
    }
  }]);
  return MenuEvent;
}();

exports["default"] = MenuEvent;
