// react-menu-list/js/Dropdown.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var Dropdown =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2["default"])(Dropdown, _React$Component);

  function Dropdown() {
    (0, _classCallCheck2["default"])(this, Dropdown);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Dropdown).apply(this, arguments));
  }

  (0, _createClass2["default"])(Dropdown, [{
    key: "render",
    value: function render() {
      return _react["default"].createElement("div", {
        style: {
          background: 'white',
          border: '1px solid rgba(0,0,0,.2)',
          boxShadow: '0 2px 4px rgba(0,0,0,0.2)',
          padding: '6px 0'
        }
      }, this.props.children);
    }
  }]);
  return Dropdown;
}(_react["default"].Component);

exports["default"] = Dropdown;
(0, _defineProperty2["default"])(Dropdown, "propTypes", {
  children: _propTypes["default"].node
});
