// react-menu-list/js/MenuItem.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ChosenEvent = _interopRequireDefault(require("./events/ChosenEvent"));

var _MenuList = require("./MenuList");

var _setRef = _interopRequireDefault(require("./lib/setRef"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var MenuItem =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2["default"])(MenuItem, _React$Component);

  function MenuItem() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, MenuItem);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(MenuItem)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_menuListHandle", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "state", {
      highlighted: false
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_el", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_elSetter", function (el) {
      _this._el = el;

      if (_this.props.domRef) {
        (0, _setRef["default"])(_this.props.domRef, el);
      }
    });
    return _this;
  }

  (0, _createClass2["default"])(MenuItem, [{
    key: "hasHighlight",
    value: function hasHighlight() {
      return this.state.highlighted;
    }
  }, {
    key: "takeKeyboard",
    value: function takeKeyboard() {
      this._menuListHandle.takeKeyboard();
    }
  }, {
    key: "releaseKeyboard",
    value: function releaseKeyboard() {
      this._menuListHandle.releaseKeyboard();
    }
  }, {
    key: "lockHighlight",
    value: function lockHighlight() {
      this._menuListHandle.lockHighlight();
    }
  }, {
    key: "unlockHighlight",
    value: function unlockHighlight() {
      this._menuListHandle.unlockHighlight();
    } // byKeyboard forces focus immediately and scrolls the item into view.
    // With it false, the highlight might be delayed depending on mouse movement
    // and won't cause anything to scroll.

  }, {
    key: "highlight",
    value: function highlight() {
      var byKeyboard = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;

      this._menuListHandle.highlight(byKeyboard);
    }
  }, {
    key: "unhighlight",
    value: function unhighlight() {
      this._menuListHandle.unhighlight();
    }
  }, {
    key: "moveCursor",
    value: function moveCursor(direction, prevCursorLocation) {
      this._menuListHandle.moveCursor(direction, prevCursorLocation);
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      var el = this._el;
      /*:: if (!el) throw new Error(); */

      this._menuListHandle = this.context.registerItem(this.props, {
        notifyHighlighted: function notifyHighlighted(highlighted, byKeyboard, direction, prevCursorLocation) {
          _this2.setState({
            highlighted: highlighted
          }, function () {
            if (highlighted && byKeyboard) {
              var _el = _this2._el;
              /*:: if (!el) throw new Error(); */

              if (typeof _el.scrollIntoViewIfNeeded === 'function') {
                _el.scrollIntoViewIfNeeded();
              } else if (_el.scrollIntoView) {
                _el.scrollIntoView();
              }
            }
          });

          if (_this2.props.onHighlightChange) {
            _this2.props.onHighlightChange(highlighted, {
              byKeyboard: byKeyboard == null ? undefined : byKeyboard,
              prevCursorLocation: prevCursorLocation == null ? undefined : prevCursorLocation,
              direction: direction == null ? undefined : direction
            });
          }
        },
        notifyEvent: function notifyEvent(event) {
          switch (event.type) {
            case 'chosen':
              /*:: if (!(event instanceof ChosenEvent)) throw new Error(); */
              if (_this2.props.onItemChosen) _this2.props.onItemChosen(event);
              break;

            case 'left':
              if (_this2.props.onLeftPushed) _this2.props.onLeftPushed(event);
              break;

            case 'right':
              if (_this2.props.onRightPushed) _this2.props.onRightPushed(event);
              break;
          }
        }
      }, el);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this._menuListHandle.unregister();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      this._menuListHandle.updateProps(this.props);
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      var _this$props = this.props,
          children = _this$props.children,
          onMouseLeave = _this$props.onMouseLeave;
      var highlighted = this.state.highlighted;
      var style = this.props.style;
      var className = this.props.className;

      if (highlighted) {
        if (this.props.highlightedStyle) {
          style = _objectSpread({}, style, {}, this.props.highlightedStyle);
        }

        if (this.props.highlightedClassName) {
          className = "".concat(className || '', " ").concat(this.props.highlightedClassName);
        }
      }

      return _react["default"].createElement("div", {
        ref: this._elSetter,
        style: style,
        className: className,
        onClick: function onClick() {
          return _this3._menuListHandle.itemChosen();
        },
        onMouseEnter: function onMouseEnter() {
          return _this3.highlight(false);
        },
        onMouseLeave: onMouseLeave || function () {
          return _this3.unhighlight();
        },
        role: "menuitem",
        "aria-haspopup": this.props['aria-haspopup'],
        "aria-expanded": this.props['aria-expanded']
      }, children);
    }
  }]);
  return MenuItem;
}(_react["default"].Component);

exports["default"] = MenuItem;
(0, _defineProperty2["default"])(MenuItem, "propTypes", {
  onItemChosen: _propTypes["default"].func,
  onHighlightChange: _propTypes["default"].func,
  onLeftPushed: _propTypes["default"].func,
  onRightPushed: _propTypes["default"].func,
  className: _propTypes["default"].string,
  style: _propTypes["default"].object,
  highlightedClassName: _propTypes["default"].string,
  highlightedStyle: _propTypes["default"].object,
  index: _propTypes["default"].number,
  onMouseLeave: _propTypes["default"].func,
  children: _propTypes["default"].node,
  domRef: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].func]),
  'aria-haspopup': _propTypes["default"].bool,
  'aria-expanded': _propTypes["default"].bool
});
(0, _defineProperty2["default"])(MenuItem, "contextType", _MenuList.MenuListContext);
