// react-menu-list/js/MenuListInspector.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.MenuListInspectorContext = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ChosenEvent = _interopRequireDefault(require("./events/ChosenEvent"));

var MenuListInspectorContext = _react["default"].createContext(null);

exports.MenuListInspectorContext = MenuListInspectorContext;

var MenuListInspector =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2["default"])(MenuListInspector, _React$Component);

  function MenuListInspector() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, MenuListInspector);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(MenuListInspector)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_descendantMenuLists", []);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_menuListInspectorContext", {
      registerMenuList: function registerMenuList(menuList) {
        _this._descendantMenuLists.push(menuList);
      },
      unregisterMenuList: function unregisterMenuList(menuList) {
        var i = _this._descendantMenuLists.indexOf(menuList);

        if (i < 0) throw new Error('MenuList not registered');

        _this._descendantMenuLists.splice(i, 1);
      },
      dispatchEvent: function dispatchEvent(event) {
        switch (event.type) {
          case 'chosen':
            /*:: if (!(event instanceof ChosenEvent)) throw new Error(); */
            if (_this.props.onItemChosen) _this.props.onItemChosen(event);
            break;

          case 'left':
            if (_this.props.onLeftPushed) _this.props.onLeftPushed(event);
            break;

          case 'right':
            if (_this.props.onRightPushed) _this.props.onRightPushed(event);
            break;
        }

        if (event.cancelBubble) return;

        var parentCtx = _this._parentCtx();

        if (parentCtx) {
          parentCtx.dispatchEvent(event);
        }
      }
    });
    return _this;
  }

  (0, _createClass2["default"])(MenuListInspector, [{
    key: "_parentCtx",
    value: function _parentCtx() {
      return this.context;
    }
  }, {
    key: "moveCursor",
    value: function moveCursor(direction, prevCursorLocation) {
      var menuList = this._descendantMenuLists[0];

      if (!menuList) {
        return false;
      }

      menuList.moveCursor(direction, prevCursorLocation);
      return true;
    }
  }, {
    key: "hasHighlight",
    value: function hasHighlight() {
      for (var i = 0, len = this._descendantMenuLists.length; i < len; i++) {
        if (this._descendantMenuLists[i].hasHighlight()) {
          return true;
        }
      }

      return false;
    }
  }, {
    key: "render",
    value: function render() {
      return _react["default"].createElement(MenuListInspectorContext.Provider, {
        value: this._menuListInspectorContext
      }, this.props.children);
    }
  }]);
  return MenuListInspector;
}(_react["default"].Component);

exports["default"] = MenuListInspector;
(0, _defineProperty2["default"])(MenuListInspector, "propTypes", {
  onItemChosen: _propTypes["default"].func,
  onLeftPushed: _propTypes["default"].func,
  onRightPushed: _propTypes["default"].func,
  children: _propTypes["default"].element
});
(0, _defineProperty2["default"])(MenuListInspector, "contextType", MenuListInspectorContext);
