// react-menu-list/js/MenuList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.MenuListContext = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _kefir = _interopRequireDefault(require("kefir"));

var _kefirStopper = _interopRequireDefault(require("kefir-stopper"));

var _arrayFindIndex = _interopRequireDefault(require("array-find-index"));

var _fromEventsCapture = _interopRequireDefault(require("./lib/fromEventsCapture"));

var _MenuEvent = _interopRequireDefault(require("./events/MenuEvent"));

var _ChosenEvent = _interopRequireDefault(require("./events/ChosenEvent"));

var _MenuListInspector = require("./MenuListInspector");

var MenuListContext = _react["default"].createContext(null);

exports.MenuListContext = MenuListContext;

var MenuList =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2["default"])(MenuList, _React$Component);

  function MenuList() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, MenuList);

    for (var _len = arguments.length, args = new Array(_len), _key2 = 0; _key2 < _len; _key2++) {
      args[_key2] = arguments[_key2];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(MenuList)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_stopper", (0, _kefirStopper["default"])());
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_listItems", []);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_naturalHighlightedIndex", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_lockedHighlightedIndex", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_keyboardTakenByIndex", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_elRef", _react["default"].createRef());
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_menuListContext", {
      registerItem: function registerItem(props, control, el) {
        var item = {
          props: props,
          control: control,
          el: el
        };

        var register = function register() {
          var i = -1;

          if (item.props.index == null) {
            i = (0, _arrayFindIndex["default"])(_this._listItems, function (_item) {
              return (item.el.compareDocumentPosition(_item.el) & Node.DOCUMENT_POSITION_PRECEDING) === 0;
            });
          } else {
            i = (0, _arrayFindIndex["default"])(_this._listItems, function (_item) {
              return _item.props.index != null && item.props.index < _item.props.index;
            });
          }

          if (i < 0) {
            _this._listItems.push(item);
          } else {
            _this._listItems.splice(i, 0, item);

            if (_this._naturalHighlightedIndex != null && i <= _this._naturalHighlightedIndex) {
              _this._naturalHighlightedIndex++;
            }

            if (_this._lockedHighlightedIndex != null && i <= _this._lockedHighlightedIndex) {
              _this._lockedHighlightedIndex++;
            }

            if (_this._keyboardTakenByIndex != null && i <= _this._keyboardTakenByIndex) {
              _this._keyboardTakenByIndex++;
            }
          }
        };

        register();
        var menuListHandle = {
          highlight: function highlight(byKeyboard) {
            var i = _this._listItems.indexOf(item);

            if (i < 0) throw new Error('Already unregistered MenuItem');

            _this._naturalHighlight(i, byKeyboard);
          },
          unhighlight: function unhighlight() {
            var i = _this._listItems.indexOf(item);

            if (i < 0) throw new Error('Already unregistered MenuItem');

            if (_this._naturalHighlightedIndex === i) {
              _this._naturalHighlight(null, false);
            }
          },
          itemChosen: function itemChosen() {
            _this._dispatchEvent(control, new _ChosenEvent["default"]('chosen', false));
          },
          takeKeyboard: function takeKeyboard() {
            var i = _this._listItems.indexOf(item);

            if (i < 0) throw new Error('Already unregistered MenuItem');
            _this._keyboardTakenByIndex = i;
          },
          releaseKeyboard: function releaseKeyboard() {
            var i = _this._listItems.indexOf(item);

            if (i < 0) throw new Error('Already unregistered MenuItem');

            if (_this._keyboardTakenByIndex === i) {
              _this._keyboardTakenByIndex = null;
            }
          },
          lockHighlight: function lockHighlight() {
            var i = _this._listItems.indexOf(item);

            if (i < 0) throw new Error('Already unregistered MenuItem');

            _this._lockHighlight(i);
          },
          unlockHighlight: function unlockHighlight() {
            var i = _this._listItems.indexOf(item);

            if (i < 0) throw new Error('Already unregistered MenuItem');

            if (_this._lockedHighlightedIndex === i) {
              _this._lockHighlight(null);
            }
          },
          moveCursor: function moveCursor(direction, prevCursorLocation) {
            _this.moveCursor(direction, prevCursorLocation);
          },
          updateProps: function updateProps(newProps) {
            if (item.props.index !== newProps.index) {
              var oldIndex = _this._listItems.indexOf(item);

              var isNaturalHighlightIndex = _this._naturalHighlightedIndex === oldIndex;
              var isLockedHighlightIndex = _this._lockedHighlightedIndex === oldIndex;
              var isKeyboardTakenByIndex = _this._keyboardTakenByIndex === oldIndex;
              menuListHandle.unregister();
              props = newProps;
              item.props = newProps;
              register();

              if (isNaturalHighlightIndex || isLockedHighlightIndex || isKeyboardTakenByIndex) {
                var newIndex = _this._listItems.indexOf(item);

                if (isNaturalHighlightIndex) _this._naturalHighlightedIndex = newIndex;
                if (isLockedHighlightIndex) _this._lockedHighlightedIndex = newIndex;
                if (isKeyboardTakenByIndex) _this._keyboardTakenByIndex = newIndex;
              }
            } else {
              props = newProps;
              item.props = newProps;
            }
          },
          unregister: function unregister() {
            var i = _this._listItems.indexOf(item);

            if (i < 0) throw new Error('Already unregistered MenuItem');

            if (i === _this._naturalHighlightedIndex) {
              _this._naturalHighlightedIndex = null;
            } else if (_this._naturalHighlightedIndex != null && i < _this._naturalHighlightedIndex) {
              _this._naturalHighlightedIndex--;
            }

            if (i === _this._lockedHighlightedIndex) {
              _this._lockedHighlightedIndex = null;
            } else if (_this._lockedHighlightedIndex != null && i < _this._lockedHighlightedIndex) {
              _this._lockedHighlightedIndex--;
            }

            if (i === _this._keyboardTakenByIndex) {
              _this._keyboardTakenByIndex = null;
            } else if (_this._keyboardTakenByIndex != null && i < _this._keyboardTakenByIndex) {
              _this._keyboardTakenByIndex--;
            }

            _this._listItems.splice(i, 1);
          }
        };
        return menuListHandle;
      }
    });
    return _this;
  }

  (0, _createClass2["default"])(MenuList, [{
    key: "_getVisibleHighlightedIndex",
    value: function _getVisibleHighlightedIndex() {
      return this._lockedHighlightedIndex != null ? this._lockedHighlightedIndex : this._naturalHighlightedIndex;
    }
  }, {
    key: "_parentCtx",
    value: function _parentCtx() {
      return this.context;
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      var isEnterOrArrowKey = function isEnterOrArrowKey(e) {
        return e.which === 13 || 37 <= e.which && e.which <= 40;
      };

      var el = this._elRef.current;
      /*:: if (!el) throw new Error(); */
      // The only things that should receive keydown/keypress events before us
      // are our children. This allows a MenuItem to contain a text input
      // which selectively stops propagation on key events for example.

      _kefir["default"].merge([_kefir["default"].fromEvents(window, 'keydown').filter(isEnterOrArrowKey).filter(function (e) {
        return el.contains(e.target);
      }), (0, _fromEventsCapture["default"])(window, 'keydown').filter(isEnterOrArrowKey).filter(function (e) {
        return !el.contains(e.target);
      })]).takeUntilBy(this._stopper).onValue(function (event) {
        return _this2._key(event);
      });

      var parentCtx = this._parentCtx();

      if (parentCtx) {
        parentCtx.registerMenuList(this);
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this._stopper.destroy();

      var parentCtx = this._parentCtx();

      if (parentCtx) {
        parentCtx.unregisterMenuList(this);
      }
    }
  }, {
    key: "_naturalHighlight",
    value: function _naturalHighlight(index, byKeyboard, direction, prevCursorLocation) {
      var visibleHighlightedIndex = this._getVisibleHighlightedIndex();

      if (this._lockedHighlightedIndex != null && byKeyboard) {
        this._lockedHighlightedIndex = null;
      }

      this._naturalHighlightedIndex = index;

      if (this._lockedHighlightedIndex == null) {
        if (index != null) {
          this._listItems[index].control.notifyHighlighted(true, byKeyboard, direction, prevCursorLocation);
        }

        if (visibleHighlightedIndex != null && visibleHighlightedIndex != index) {
          this._listItems[visibleHighlightedIndex].control.notifyHighlighted(false);
        }
      }
    }
  }, {
    key: "_lockHighlight",
    value: function _lockHighlight(index) {
      if (index === this._lockedHighlightedIndex) return;

      var visibleHighlightedIndex = this._getVisibleHighlightedIndex();

      this._lockedHighlightedIndex = index;

      var newVisibleHighlightedIndex = this._getVisibleHighlightedIndex();

      if (visibleHighlightedIndex != null && newVisibleHighlightedIndex == null) {
        // When unlocking, prefer to keep the current selection over de-selecting
        // everything.
        this._naturalHighlightedIndex = visibleHighlightedIndex;
      } else if (visibleHighlightedIndex != newVisibleHighlightedIndex) {
        if (visibleHighlightedIndex != null) {
          this._listItems[visibleHighlightedIndex].control.notifyHighlighted(false);
        }

        if (newVisibleHighlightedIndex != null) {
          this._listItems[newVisibleHighlightedIndex].control.notifyHighlighted(true, false);
        } else if (this._naturalHighlightedIndex != null) {
          this._listItems[this._naturalHighlightedIndex].control.notifyHighlighted(true, false);
        }
      }
    }
  }, {
    key: "_dispatchEvent",
    value: function _dispatchEvent(control, event) {
      if (control) {
        control.notifyEvent(event);
        if (event.cancelBubble) return;
      }

      switch (event.type) {
        case 'chosen':
          /*:: if (!(event instanceof ChosenEvent)) throw new Error(); */
          if (this.props.onItemChosen) this.props.onItemChosen(event);
          break;
        // case 'up':
        //   break;
        // case 'down':
        //   break;

        case 'left':
          if (this.props.onLeftPushed) this.props.onLeftPushed(event);
          break;

        case 'right':
          if (this.props.onRightPushed) this.props.onRightPushed(event);
          break;
      }

      if (event.cancelBubble) return;

      var parentCtx = this._parentCtx();

      if (parentCtx) {
        parentCtx.dispatchEvent(event);
      }
    }
  }, {
    key: "_key",
    value: function _key(event) {
      if (this._keyboardTakenByIndex != null || this._listItems.length === 0) {
        return;
      }

      var visibleHighlightedIndex = this._getVisibleHighlightedIndex(); // enter, left, right activate for the current visibly selected item.
      // up and down de-activate any locks in place, so that they act from the last
      // naturally-selected item.


      var mEvent = null;

      switch (event.which) {
        case 13:
          //enter
          if (visibleHighlightedIndex != null) {
            mEvent = new _ChosenEvent["default"]('chosen', true);
            event.preventDefault();
            event.stopPropagation();
          }

          break;

        case 37:
          //left
          if (visibleHighlightedIndex != null) {
            mEvent = new _MenuEvent["default"]('left');
          }

          break;

        case 39:
          //right
          if (visibleHighlightedIndex != null) {
            mEvent = new _MenuEvent["default"]('right');
          }

          break;

        case 38:
          //up
          event.preventDefault();
          event.stopPropagation();
          this.moveCursor('up');
          break;

        case 40:
          //down
          event.preventDefault();
          event.stopPropagation();
          this.moveCursor('down');
          break;
      }

      if (mEvent) {
        var _control = visibleHighlightedIndex == null ? null : this._listItems[visibleHighlightedIndex].control;

        this._dispatchEvent(_control, mEvent);

        if (mEvent.defaultPrevented) event.preventDefault();
        if (mEvent.defaultPrevented || mEvent.cancelBubble) event.stopPropagation();
      }
    }
  }, {
    key: "moveCursor",
    value: function moveCursor(direction, prevCursorLocation) {
      if (this._listItems.length == 0) return;

      switch (direction) {
        case 'up':
          if (this._naturalHighlightedIndex == null || this._naturalHighlightedIndex == 0) {
            this._naturalHighlight(this._listItems.length - 1, true, direction, prevCursorLocation);
          } else {
            this._naturalHighlight(this._naturalHighlightedIndex - 1, true, direction, prevCursorLocation);
          }

          break;

        case 'down':
          if (this._naturalHighlightedIndex == null || this._naturalHighlightedIndex == this._listItems.length - 1) {
            this._naturalHighlight(0, true, direction, prevCursorLocation);
          } else {
            this._naturalHighlight(this._naturalHighlightedIndex + 1, true, direction, prevCursorLocation);
          }

          break;
      }
    }
  }, {
    key: "hasHighlight",
    value: function hasHighlight() {
      return this._getVisibleHighlightedIndex() != null;
    }
  }, {
    key: "render",
    value: function render() {
      return _react["default"].createElement("div", {
        role: "menu",
        ref: this._elRef
      }, _react["default"].createElement(MenuListContext.Provider, {
        value: this._menuListContext
      }, this.props.children));
    }
  }]);
  return MenuList;
}(_react["default"].Component);

exports["default"] = MenuList;
(0, _defineProperty2["default"])(MenuList, "propTypes", {
  onItemChosen: _propTypes["default"].func,
  onLeftPushed: _propTypes["default"].func,
  onRightPushed: _propTypes["default"].func,
  children: _propTypes["default"].node
});
(0, _defineProperty2["default"])(MenuList, "contextType", _MenuListInspector.MenuListInspectorContext);
